File: solutions/[124]二叉树中的最大路径和.cpp
#include <cstdlib>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
	int maxPathSum(TreeNode* root) {
		int ans = -99999;
		dp(root, ans);
		return ans;
	}
	int dp(TreeNode* root, int& ans) {
		if (root == nullptr)
			return 0;

		if (root->val > ans)
			ans = root->val;
		if (root->left == nullptr && root->right == nullptr) {
			return root->val;
		}


		int left = dp(root->left, ans);
		int right = dp(root->right, ans);
		int ret = (left > 0 || right > 0 ? (left > right ? left : right) : 0) + root->val;
		int tmp = (left > 0 && right > 0 ? left + right + root->val : ret);
		if (tmp > ans)
			ans = tmp;
		return ret;
	}
};

File: solutions/[124]二叉树中的最大路径和.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        def helper(root):
            if root is None:
                return 0, -float('inf')
            
            l_length, l_ans = helper(root.left)
            r_length, r_ans = helper(root.right)

            return (max(0, l_length, r_length) + root.val,
                max(l_ans, r_ans, max(0, l_length) + max(0, r_length) + root.val))
        return helper(root)[1]
